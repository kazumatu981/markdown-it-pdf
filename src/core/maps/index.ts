export * from './contents-map';
export * from './extension-map';
export * from './render-map';
export * from './contents-manager';
