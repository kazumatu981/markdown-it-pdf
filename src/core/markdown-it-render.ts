import MarkdownIt from 'markdown-it';
import fsPromises from 'fs/promises';

export interface Styles {
    internalUrls: string[];
    externalUrls: string[];
}

export class MarkdownItRender {
    private _md: MarkdownIt = new MarkdownIt();
    private _styles: Styles = {
        internalUrls: [],
        externalUrls: [],
    };

    /**
     * Adds the given URLs to the list of internal styles for this instance.
     *
     * @param {string[]} urls - The URLs of the styles to add.
     * @return {this} - Returns the current instance for method chaining.
     */
    public addStyles(urls: string[]): this {
        this._styles.internalUrls = urls;
        return this;
    }

    /**
     * Sets the external style URLs of this instance.
     *
     * @param {string[]} urls - The URLs to set as external styles.
     * @return {this} - Returns the current instance for method chaining.
     */
    public addExternalStyles(urls: string[]): this {
        this._styles.externalUrls = urls;
        return this;
    }

    /**
     * Uses the provided plugin with optional parameters.
     *
     * @param {MarkdownIt.PluginWithParams} plugin - The plugin to use.
     * @param {any[]} params - Optional parameters for the plugin.
     * @return {this} The MarkdownItRender instance for chaining.
     */
    public use(plugin: MarkdownIt.PluginWithParams, ...params: any[]): this {
        this._md.use(plugin, ...params);
        return this;
    }

    /**
     * Renders the given markdown string into HTML.
     *
     * @param {string} markdown - The markdown string to render.
     * @return {string} The rendered HTML string.
     */
    public render(markdown: string): string {
        const htmlBody = this._md.render(markdown);
        const styleTags = MarkdownItRender.generateStyleTags(
            this._styles.internalUrls,
            this._styles.externalUrls
        );
        return MarkdownItRender.htmlTemplate
            .replace(MarkdownItRender.styles, styleTags)
            .replace(MarkdownItRender.contents, htmlBody);
    }
    /**
     * Asynchronously renders a markdown file to HTML.
     *
     * @param {string} markdownFilePath - The path to the markdown file.
     * @return {Promise<string>} A promise that resolves to the rendered HTML.
     */
    public async renderFromFileAsync(
        markdownFilePath: string
    ): Promise<string> {
        const markdown = await fsPromises.readFile(markdownFilePath, 'utf8');
        return this.render(markdown);
    }

    //#region private members
    //#region Static Variables
    private static styles = '<!-- [[[STYLES]]] -->';
    private static contents = '<!-- [[[CONTENTS]]] -->';

    private static htmlTemplate = `
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Document</title>
    ${MarkdownItRender.styles}
</head>
<body>
    ${MarkdownItRender.contents}
</body>`;
    //#endregion

    //#region Static Methods
    /**
     * Generates style tag based on the given URL.
     * @param {string} stylePath css file path
     * @returns {string} generated style tag
     */
    private static generateStyleTag(stylePath: string): string {
        return `<link rel="stylesheet" type="text/css" href="${stylePath}">`;
    }
    /**
     * Generates style tags based on the given internal and external URLs.
     *
     * @param {string[]} internalUrls - An optional array of internal URLs.
     * @param {string[]} externalUrls - An optional array of external URLs.
     * @return {string} The generated style tags as a string.
     */
    private static generateStyleTags(
        internalUrls?: string[],
        externalUrls?: string[]
    ): string {
        return (
            (internalUrls?.map(MarkdownItRender.generateStyleTag).join('\n') ??
                '') +
            (externalUrls?.map(MarkdownItRender.generateStyleTag).join('\n') ??
                '')
        );
    }
    //#endregion
    //#endregion
}
