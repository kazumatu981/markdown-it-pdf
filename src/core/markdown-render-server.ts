import http from 'http';
import {
    type ContentsMapOptions,
    type RenderedEntity,
    RenderMap,
    ContentsMap,
} from './maps';
import {
    MarkdownItRender,
    utf8PlainTextRender,
    type HljsConfig,
} from './render';
import { type ListeningOptions, tryToListen } from './utils';
import { type Logger } from '../common';

/**
 * The options for the render server
 */
export interface RenderServerOptions
    extends ContentsMapOptions,
        ListeningOptions {
    /**
     * The port to listen on
     */
    port?: number;

    /**
     * The external urls for styles
     */
    externalUrls?: string[];
    /**
     * The template html file path
     */
    templatePath?: string;
    /**
     * The highlight.js config
     */
    hljs?: HljsConfig | false;
}

// TODO support user custom plugins on options.

/**
 * The markdown render server
 */
export class MarkdownRenderServer extends MarkdownItRender {
    private _options?: RenderServerOptions;
    private _server?: http.Server;
    private _contentsMap?: ContentsMap;
    private _listeningPort?: number;

    /**
     * Create a new markdown render server
     * @param rootDir {string} path to root directory of markdown files
     * @param options {RenderServerOptions} the options on this server launch.
     * @param logger {Logger} the logger
     * @returns {Promise<MarkdownRenderServer>} the markdown render server
     */
    public static async createInstance(
        rootDir?: string,
        options?: RenderServerOptions,
        logger?: Logger
    ): Promise<MarkdownRenderServer> {
        logger?.debug(
            `MarkdownRenderServer.createRenderServer() called with options: ${JSON.stringify(options)}`
        );

        // create this instance
        const theInstance = await new MarkdownRenderServer()
            // set the options
            .setOptions(options)
            // set the logger
            .setLogger(logger)
            // configure the template
            .configureTemplate({ ...options })
            // configure the contents map
            .then((instance) => {
                return instance.configureContentsMap(rootDir, options);
            })
            // refresh the contents map
            .then((instance) => {
                return instance.refresh(false);
            });
        logger?.debug(
            'contentsUrls: %o',
            theInstance.contentsMap.getEntityPaths()
        );

        return theInstance;
    }

    /**
     * Markdown paths which is available.
     * @returns {string[]} Markdown paths
     */
    public get availableMarkdownPaths(): string[] {
        return this.contentsMap.getEntityPaths('markdown');
    }

    /**
     * Style paths which is available.
     * @returns {string[]} Style paths
     */
    public get availableStylePaths(): string[] {
        return this.contentsMap.getEntityPaths('style');
    }

    /**
     * The url of the server. `http://localhost:{port}`
     * @returns {string} the url
     */
    public get myUrl(): string {
        return `http://localhost:${this._listeningPort}`;
    }

    private set contentsMap(contentsMap: ContentsMap) {
        this._contentsMap = contentsMap;
    }

    /**
     * The contents map
     * @returns {ContentsMap} the contents map
     */
    public get contentsMap(): ContentsMap {
        return this._contentsMap as ContentsMap;
    }

    /**
     * The listening port
     * @returns {number} the listening port
     */
    public get listeningPort(): number | undefined {
        return this._listeningPort;
    }

    /**
     * Start to listen from clients request.
     * @param port {number | undefined} The port to listen on
     * @returns {Promise<number>} A promise that resolves when the server is listening
     */
    public async listen(port?: number): Promise<number> {
        const portCandidate = port ?? this._options?.port;
        const serverPort = await tryToListen(
            portCandidate,
            this._options,
            this._logger
        );
        if (serverPort === undefined) {
            this._logger?.error('Can not listen on port %d', portCandidate);
            this._logger?.trace('Port not available');
            throw new Error('Can not listen');
        }
        this._listeningPort = serverPort.port;
        this._server = serverPort.httpServer;

        this._logger?.debug('Listening on port %d', this._listeningPort);

        // bind the server event listener
        this._server.on('request', this.onRequest.bind(this));
        return this._listeningPort;
    }

    /**
     * Close the server
     * @returns {Promise<void>} A promise that resolves when the server is closed
     */
    public async close(): Promise<void> {
        if (this._server === undefined) {
            return Promise.resolve();
        }
        this._listeningPort = undefined;
        return new Promise((resolve, rejects) => {
            (this._server as http.Server).close((err) => {
                if (err) {
                    this._logger?.error('Can not close server');
                    this._logger?.trace(err);
                    rejects(err);
                } else {
                    resolve();
                }
            });
        });
    }

    /**
     * Refresh the contents of the server
     * @param refreshContentsMap {boolean} whether to refresh the contents map
     * @returns {this} this instance, for method chaining
     */
    public async refresh(refreshContentsMap: boolean = true): Promise<this> {
        if (refreshContentsMap) {
            await this.contentsMap.refresh();
        }
        this.clearStyles();
        this.addStyles(this.availableStylePaths);
        this.clearExternalStyles();
        this.addExternalStyles(this._options?.externalUrls ?? []);
        return this;
    }

    //#region private methods

    //#region HTTP responses
    /**
     * Listener for incoming requests
     * @param req Incoming message
     * @param res Server response
     */
    private onRequest(
        req: http.IncomingMessage,
        res: http.ServerResponse
    ): void {
        this._logger?.debug('Request: %s', req.url);
        // Get requested file path
        const filePath = req.url as string;
        // Render the file contents with given path
        (this._contentsMap as ContentsMap).render(filePath).then((entity) => {
            // If file not found or rendering failed
            // Return "Not Found" to the client
            if (!entity) {
                this.writeNotFound(res);
                return;
            }
            // Write the rendered entity to the response
            this.writeEntity(res, entity);
        });
    }

    private writeNotFound(res: http.ServerResponse): void {
        this._logger?.warn(
            'Response for url: %s, status code: %d',
            res.req.url,
            404
        );
        res.statusCode = 404;
        res.appendHeader('Content-Type', 'text/plain');
        res.write('Not Found');
        res.end();
    }
    private writeEntity(
        res: http.ServerResponse,
        entity: RenderedEntity
    ): void {
        this._logger?.debug(
            'Response for url: %s, status code: %d',
            res.req.url,
            200
        );
        this._logger?.debug('content type: %s', entity.contentType);

        res.statusCode = 200;
        res.appendHeader('Content-Type', entity.contentType);
        res.write(entity.contents);
        res.end();
    }
    //#endregion

    //#region configuration functions
    private setOptions(options?: RenderServerOptions): this {
        this._options = options;
        return this;
    }
    private setLogger(logger?: Logger): this {
        this._logger = logger;
        return this;
    }
    private async configureContentsMap(
        rootDir?: string,
        options?: RenderServerOptions
    ): Promise<this> {
        const contentMap = await ContentsMap.createInstance(
            new RenderMap(),
            rootDir,
            options
        );
        contentMap.renderMap
            .set('markdown', this)
            .set('style', utf8PlainTextRender)
            .set('plainText', utf8PlainTextRender);
        this.contentsMap = contentMap;

        return this;
    }

    //#endregion

    //#endregion
}
