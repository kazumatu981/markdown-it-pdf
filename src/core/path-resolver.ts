import fs from 'fs';
import path from 'path';
import fsPromises from 'fs/promises';

type IsMatchFunction = (filePath: string) => boolean;
type TransformFunction<T> = (filePath: string) => T;

/**
 * Finds files in a given directory and returns an array of transformed results.
 *
 * @param {string} rootDir - The root directory to start the search from.
 * @param {boolean} recursive - Whether to search recursively in subdirectories.
 * @param {IsMatchFunction} isMatch - A function to check if a file matches a certain criteria.
 * @param {TransformFunction<T>} transform - A function to transform each found file.
 * @return {Promise<T[]>} A promise that resolves to an array of transformed file results.
 */
export async function findFiles<T>(
    rootDir: string,
    recursive: boolean,
    isMatch: IsMatchFunction,
    transform: TransformFunction<T>
): Promise<T[]> {
    const elements = await fsPromises.readdir(rootDir, { withFileTypes: true });
    const fileElements = elements
        .filter((e) => e.isFile() && isMatch(e.name))
        .map((e) => path.join(rootDir, e.name))
        .map(transform);
    if (recursive) {
        const subFileElements = await Promise.all(
            elements
                .filter((e) => e.isDirectory())
                .map((e) => path.join(rootDir, e.name))
                .map((d) => findFiles(d, recursive, isMatch, transform))
        );

        fileElements.push(...subFileElements.flat());
    }
    return fileElements;
}
/**
 * Converts a file path to a URL relative to the root directory.
 *
 * @param {string} rootDir - The root directory path.
 * @param {string} filePath - The file path to convert.
 * @return {string} The URL representing the file path relative to the root directory.
 * @throws {Error} If the file path is not in the root directory.
 */
export function filePathToUrl(rootDir: string, filePath: string): string {
    rootDir = path.resolve(process.cwd(), rootDir);
    filePath = path.resolve(process.cwd(), filePath);
    const relativePath = normalizePath(path.relative(rootDir, filePath));
    if (relativePath.startsWith('.')) {
        throw new Error('Path is not in rootDir');
    }
    return `/${relativePath}`;
}

/**
 * Normalizes a file path by converting it to a relative path and replacing any
 * Windows-style path separators with POSIX-style separators. Throws an error if
 * the file path is absolute.
 *
 * @param {string} filePath - The file path to normalize.
 * @return {string} The normalized file path.
 * @throws {Error} If the file path is absolute.
 */
export function normalizePath(filePath: string): string {
    if (path.isAbsolute(filePath)) {
        throw new Error('Path must be relative');
    }
    return path.normalize(filePath).split(path.win32.sep).join(path.posix.sep);
}

/**
 * Asynchronously builds a tree of directories based on the provided array of file paths.
 *
 * @param {string[]} files - An array of file paths.
 * @return {Promise<void>} A promise that resolves when the tree of directories is built successfully, or rejects with an error if there was a problem.
 */
export async function buildTreeOfFiles(files: string[]): Promise<void> {
    const dirs = Array.from(new Set(files.map((file) => path.dirname(file))));
    return new Promise((resolve, reject) => {
        Promise.all<void>(
            dirs.map(async (dir) => {
                await fsPromises.mkdir(dir, { recursive: true });
            })
        )
            .then((_) => resolve())
            .catch((err) => reject(err));
    });
}
