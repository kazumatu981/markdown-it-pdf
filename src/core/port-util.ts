import http from 'http';

interface Range {
    min?: number;
    max?: number;
}
interface SafeRange {
    min: number;
    max: number;
}
const privatePortRange: SafeRange = {
    min: 49152,
    max: 65535,
};

export interface ServerPort {
    port: number;
    httpServer: http.Server;
}

export interface ServerPortOptions {
    retry?: number;
    range?: Range;
}

function getRandom(range?: Range): number {
    const candidate = {
        min: range?.min ?? privatePortRange.min,
        max: range?.max ?? privatePortRange.max,
    };
    if (candidate.min > candidate.max) {
        throw new Error('Invalid range');
    }
    return Math.floor(
        Math.random() * (candidate.max - candidate.min) + candidate.min
    );
}

export async function tryToListen(
    port?: number,
    options?: ServerPortOptions
): Promise<ServerPort | undefined> {
    let serverPort = undefined;
    if (port != undefined) {
        serverPort = await tryToListenCore(port);
    } else {
        for (let i = 0; i < (options?.retry ?? 10); i++) {
            port = getRandom(options?.range);
            serverPort = await tryToListenCore(port);
            if (serverPort !== undefined) {
                break; // success
            }
        }
    }
    return serverPort;
}
function tryToListenCore(port: number): Promise<ServerPort | undefined> {
    const server = http.createServer();
    return new Promise((resolve) => {
        server.listen(port, () => {
            resolve({
                port,
                httpServer: server,
            });
        });
        server.on('error', () => {
            resolve(undefined);
        });
    });
}
