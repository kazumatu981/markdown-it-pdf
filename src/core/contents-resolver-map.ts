import { MarkdownItRender } from './markdown-it-render';
import fsPromises from 'fs/promises';
export type ContentType = 'markdown' | 'style';
export type ContentsResolverFunction = (filePath: string) => Promise<string>;

export class ContentsResolver extends Map<
    ContentType,
    ContentsResolverFunction
> {
    private _render: MarkdownItRender;
    public constructor(render: MarkdownItRender) {
        super();
        this._render = render;
        this.set('markdown', async (filePath) => {
            return this._render.renderFromFileAsync(filePath);
        });
        this.set('style', async (filePath) => {
            return fsPromises.readFile(filePath, 'utf8');
        });
    }

    public static resolveMediaType(contentType: ContentType): string{
        switch (contentType) {
            case 'markdown':
                return 'text/html';
            case 'style':
                return 'text/css';
            default:
                return 'text/plain';
        }
    }
    public async resolve(type: ContentType, filePath: string): Promise<string> {
        const resolver = this.get(type);
        if (!resolver) {
            throw new Error(`Unsupported contents type: ${type}`);
        }
        return resolver(filePath);
    }
}
