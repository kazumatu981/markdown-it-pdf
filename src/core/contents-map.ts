import { type MarkdownItRender } from './markdown-it-render';
import { findFiles, filePathToUrl } from './path-resolver';
import { type ResolverType, isSupported, getResolver } from './resolver-map';
import fsPromises from 'fs/promises';
import path from 'path';
export type ContentsResolverFunction = (
    filePath: string
) => Promise<string | Buffer>;

export interface ContentsResolver {
    mediaType: string;
    resolve: ContentsResolverFunction;
}

export interface ContentsMapEntity {
    url: string;
    resolverType: ResolverType;
    contentType: string;
    contentPath: string;
}

export interface RenderedEntity extends ContentsMapEntity {
    contents: string | Buffer;
}

export interface ContentsMapOptions {
    recursive?: boolean;
    externalUrls?: string[];
}
export class ContentsMap extends Map<string, ContentsMapEntity> {
    private _resolver: Map<ResolverType, ContentsResolverFunction>;
    private _render: MarkdownItRender;

    public static async createInstance(
        render: MarkdownItRender,
        contentsRoot: string,
        options?: ContentsMapOptions
    ): Promise<ContentsMap> {
        // create the instance
        const theInstance = new ContentsMap(render);

        // resolve to entity
        const contents = await findFiles<ContentsMapEntity>(
            contentsRoot,
            options?.recursive ?? true,
            isSupported,
            (filePath) => {
                const resolver = getResolver(path.extname(filePath));
                return {
                    url: filePathToUrl(contentsRoot, filePath),
                    resolverType: resolver.resolverType,
                    contentType:
                        resolver.resolvedContentType ?? resolver.contentType,
                    contentPath: filePath,
                };
            }
        );

        // set into entity
        contents.forEach((entry) => {
            theInstance.set(entry.url, entry);
        });

        // add into internalUrls
        const internalUrls = contents
            .filter((entry) => entry.contentType === 'text/css')
            .map((kv) => kv.url);
        render.addStyles(internalUrls);
        // add into externalUrls
        render.addExternalStyles(options?.externalUrls ?? []);

        return theInstance;
    }

    private constructor(render: MarkdownItRender) {
        super();
        this._render = render;
        this._resolver = this.generateDefaultResolvers();
    }

    public get markdownEntryUrls(): string[] {
        return [...this.keys()].filter(
            (url) => this.get(url)?.resolverType === 'markdown'
        );
    }

    public async render(filePath: string): Promise<RenderedEntity | undefined> {
        const entity = this.get(filePath);
        if (!entity) {
            return;
        }
        const contentResolver = this._resolver.get(entity.resolverType);
        if (!contentResolver) {
            return;
        }
        const contents = await contentResolver(entity.contentPath);
        return { ...entity, contents };
    }

    //#region private methods
    private generateDefaultResolvers(): Map<
        ResolverType,
        ContentsResolverFunction
    > {
        // add default resolvers
        return new Map<ResolverType, ContentsResolverFunction>()
            .set('markdown', this._markdownResolver.bind(this))
            .set('style', this._plainTextResolver.bind(this))
            .set('plainText', this._plainTextResolver.bind(this))
            .set('binary', this._binaryResolver.bind(this));
    }
    private async _markdownResolver(filePath: string): Promise<string> {
        return this._render.renderFromFileAsync(filePath);
    }
    private async _plainTextResolver(filePath: string): Promise<string> {
        return fsPromises.readFile(filePath, 'utf8');
    }
    private async _binaryResolver(filePath: string): Promise<Buffer> {
        return fsPromises.readFile(filePath);
    }
    //#endregion
}
