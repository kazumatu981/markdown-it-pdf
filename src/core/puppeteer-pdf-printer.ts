import puppeteer from 'puppeteer';
import type { Browser, Page } from 'puppeteer';
import { buildTreeOfFiles } from './utils/path-resolver';
import path from 'path';
import { type PDFOptions } from 'puppeteer';
import { Logger } from '../common/logger';

//#region constants
export const defaultPrinterOption: PDFOptions = {
    format: 'A4',
    margin: {
        top: '12.7mm',
        bottom: '12.7mm',
        left: '12.7mm',
        right: '12.7mm',
    },
};
//#endregion
//#region types and interfaces
export type PrinterOptions = Omit<PDFOptions, 'path'>;

interface BrowserAndPage {
    browser: Browser;
    page: Page;
}
//#endregion

//#region main classes

/**
 * Base class fot the PDF printer
 */
export abstract class PuppeteerPDFPrinter<
    T = string | string[],
    U = void | Buffer,
> {
    /**
     * The base URL of the site to print.
     * @type {string}
     */
    protected _siteUrl: string;
    /**
     * The directory where the PDFs will be saved.
     * @type {string | undefined}
     */
    protected _outputDir?: string;
    /**
     * The options to configure the printer.
     * @type {PrinterOptions | undefined}
     */
    protected _options?: PrinterOptions;
    /**
     * The logger to use.
     * @type {Logger | undefined}
     */
    protected _logger?: Logger;

    /**
     * Constructor for the PuppeteerPDFPrinter.
     *
     * @param {string} siteUrl - The base URL of the site to print.
     * @param {string} [outputDir] - The directory where the PDFs will be saved.
     * @param {PrinterOptions} [options] - The options to configure the printer.
     * @param {Logger} [logger] - The logger to use.
     */
    protected constructor(
        siteUrl: string,
        outputDir?: string,
        options?: PrinterOptions,
        logger?: Logger
    ) {
        this._siteUrl = siteUrl;
        this._outputDir = outputDir;
        this._options = options;
        this._logger = logger;
    }

    /**
     * Creates a new Printer that prints the PDFs into files.
     *
     * @param {string} siteUrl - The base URL of the site to print.
     * @param {string} outputDir - The directory where the PDFs will be saved.
     * @param {PrinterOptions} [options] - The options to configure the printer.
     * @param {Logger} [logger] - The logger to use.
     * @return {FilePrinter} The newly created Printer.
     */
    public static intoFiles(
        siteUrl: string,
        outputDir: string,
        options?: PrinterOptions,
        logger?: Logger
    ): FilePrinter {
        // Create a new FilePrinter with the provided arguments
        return new FilePrinter(siteUrl, outputDir, options, logger);
    }

    /**
     * Creates a new Printer that prints the PDFs into memory.
     *
     * @param {string} siteUrl - The base URL of the site to print.
     * @param {PrinterOptions} [options] - The options to configure the printer.
     * @param {Logger} [logger] - The logger to use.
     * @return {MemoryPrinter} The newly created Printer.
     */
    public static intoMemory(
        siteUrl: string,
        options?: PrinterOptions,
        logger?: Logger
    ): MemoryPrinter {
        // Create a new Printer that prints the PDFs into memory.
        // The siteUrl parameter is the base URL of the site to print.
        // The options parameter is an object that allows you to configure the printer.
        // The logger parameter is an optional logger.
        return new MemoryPrinter(siteUrl, undefined, options, logger);
    }

    /**
     * Printer must be implemented.
     * @param targetUrls - The URLs of the pages to print.
     */
    public abstract print(targetUrls: T): Promise<U>;
}

class VirtualBrowser {
    /**
     * The Puppeteer browser instance.
     * @type {Browser}
     * @private
     */
    private _browser: Browser;

    /**
     * The Puppeteer page instance.
     * @type {Page}
     * @private
     */
    private _page: Page;

    /**
     * Creates a new instance of VirtualBrowser.
     * @param {Browser} browser - The Puppeteer browser instance.
     * @param {Page} page - The Puppeteer page instance.
     */
    protected constructor(browser: Browser, page: Page) {
        this._browser = browser;
        this._page = page;
    }

    /**
     * Launches a new instance of Puppeteer and creates a new page.
     *
     * @returns {Promise<VirtualBrowser>} A new instance of VirtualBrowser.
     */
    public static async launch(): Promise<VirtualBrowser> {
        // Launch a new instance of Puppeteer.
        const browser = await puppeteer.launch();

        // Create a new page in the browser.
        const page = await browser.newPage();

        // Return a new instance of VirtualBrowser with the launched browser and page.
        return new VirtualBrowser(browser, page);
    }

    /**
     * Navigates to the specified URL and waits for the body to load.
     *
     * @param {string} url - The URL to navigate to.
     * @returns {Promise<void>} A promise that resolves when the page is fully loaded.
     */
    public async gotoAndWait(url: string): Promise<void> {
        // Navigate to the specified URL.
        await this._page.goto(url);

        // Wait for the body to load.
        // The waitForSelector method returns a promise that resolves when the selector is found in the page.
        // In this case, we are waiting for the body selector to be found, indicating that the page is fully loaded.
        await this._page.waitForSelector('body');
    }

    /**
     * Prints the page as a PDF and returns the result.
     *
     * @param {string} filePath - The path to save the PDF to. If not specified, the PDF is returned as a buffer.
     * @param {PrinterOptions} options - Additional options for the PDF.
     * @returns {Promise<Buffer | void>} The PDF as a buffer, or void if filePath is specified.
     */
    public async print<T = Buffer | void>(
        filePath?: string,
        options?: PrinterOptions
    ): Promise<T> {
        // Use Puppeteer's page.pdf method to print the page as a PDF.
        // If filePath is specified, the PDF is saved to that file. Otherwise, the PDF is returned as a buffer.
        // The options parameter is an object that allows you to configure the PDF.
        return this._page.pdf({
            path: filePath,
            ...defaultPrinterOption,
            ...options,
        }) as Promise<T>;
    }

    /**
     * Closes the page and browser associated with the virtual browser.
     *
     * @returns {Promise<void>} A promise that resolves when the page and browser are closed.
     */
    public async close(): Promise<void> {
        // Close the page.
        await this._page.close();

        // Close the browser.
        await this._browser.close();
    }
}

/**
 * Printer that prints multiple pages into files.
 */
class FilePrinter extends PuppeteerPDFPrinter<string[], void> {
    /**
     * Prints multiple pages into files.
     *
     * @param {string[]} targetUrls - The URLs of the pages to print.
     * @return {Promise<void>} A promise that resolves when all pages are printed.
     */
    public async print(targetUrls: string[]): Promise<void> {
        // Build folder tree for the PDF files.
        await buildTreeOfFiles(
            targetUrls.map((page) => path.join(this._outputDir as string, page))
        );

        // Resolve URLs of the pages to print.
        const urls = targetUrls.map((page) => ({
            pathToPdf: `${path.join(this._outputDir as string, page)}.pdf`,
            fullUrl: `${this._siteUrl}${page}`,
        }));

        // Launch Puppeteer and create a new page.
        const browser = await VirtualBrowser.launch();
        this._logger?.debug(`Launched Puppeteer.`);

        // For each page, print it and save it as a PDF.
        for (const url of urls) {
            await browser.gotoAndWait(url.fullUrl);
            await browser.print<void>(url.pathToPdf, this._options);
            this._logger?.debug(`Printed ${url.fullUrl} to ${url.pathToPdf}.`);
        }

        // Close the page and browser.
        await browser.close();
        this._logger?.debug(`Closed Puppeteer.`);
    }
}

/**
 * Printer that prints a single page into memory.
 */
class MemoryPrinter extends PuppeteerPDFPrinter<string, Buffer> {
    /**
     * Prints a single page into memory.
     *
     * @param {string} targetUrl - The URL of the page to print.
     * @return {Promise<Buffer>} The PDF as a Buffer.
     */
    public async print(targetUrl: string): Promise<Buffer> {
        // Resolve the URL
        const fullUrl = `${this._siteUrl}${targetUrl}`;

        // Launch Puppeteer
        const browser = await VirtualBrowser.launch();
        this._logger?.debug(`Launched Puppeteer.`);

        // Print the page
        await browser.gotoAndWait(fullUrl);
        const pdf = await browser.print<Buffer>(undefined, this._options);
        this._logger?.debug(`Printed ${fullUrl} to memory.`);

        // Close the page and browser
        await browser.close();
        this._logger?.debug(`Closed Puppeteer.`);

        // Return the PDF
        return pdf;
    }
}

//#endregion
