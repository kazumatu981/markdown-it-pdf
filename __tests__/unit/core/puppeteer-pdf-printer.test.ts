import { jest, describe, it, expect } from '@jest/globals';
import puppeteer, { Browser } from 'puppeteer';
import { PuppeteerPDFPrinter } from '../../../src/core/puppeteer-pdf-printer';

jest.mock('puppeteer');

// TODO add test intoMemory with/without logger.

describe('CoreLibrary Unit Tests - PuppeteerPDFPrinter', () => {
    const gotoFn = jest.fn();
    const waitForSelectorFn = jest.fn();
    const pdfFn = jest.fn();
    const pageCloseFn = jest.fn();
    const browserCloseFn = jest.fn();
    beforeEach(() => {
        (puppeteer.launch as jest.Mock).mockResolvedValue({
            newPage: jest.fn().mockResolvedValue({
                goto: gotoFn,
                waitForSelector: waitForSelectorFn,
                pdf: pdfFn,
                close: pageCloseFn,
            } as never),
            close: browserCloseFn,
        } as never);
    });
    afterEach(() => {
        jest.clearAllMocks();
    });

    it('print into files', async () => {
        await PuppeteerPDFPrinter.intoFiles('http://localhost:3000', 'test', {
            format: 'a4',
        }).print(['/test1.md', '/test2.md']);
        expect(puppeteer.launch).toBeCalledTimes(1);
        expect(gotoFn).toBeCalledTimes(2);
        expect(gotoFn).toMatchSnapshot();
        expect(pdfFn).toBeCalledTimes(2);
        expect(pdfFn).toMatchSnapshot();
    });
    it('print into memory', async () => {
        const pdf = await PuppeteerPDFPrinter.intoMemory(
            'http://localhost:3000',
            { format: 'a4' }
        ).print('/test1.md');
        expect(gotoFn).toBeCalledTimes(1);
        expect(gotoFn).toMatchSnapshot();
        expect(pdfFn).toBeCalledTimes(1);
        expect(pdfFn).toMatchSnapshot();
    });
});
