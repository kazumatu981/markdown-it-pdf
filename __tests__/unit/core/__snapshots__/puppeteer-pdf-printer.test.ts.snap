// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoreLibrary Unit Tests - PuppeteerPDFPrinter printIntoMemory 1`] = `
[MockFunction] {
  "calls": [
    [
      "http://localhost:3000//test1.md",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`CoreLibrary Unit Tests - PuppeteerPDFPrinter printIntoMemory 2`] = `
[MockFunction] {
  "calls": [
    [
      {
        "format": "a4",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`CoreLibrary Unit Tests - PuppeteerPDFPrinter printManyPages 1`] = `
[MockFunction] {
  "calls": [
    [
      "http://localhost:3000/test1.md",
    ],
    [
      "http://localhost:3000/test2.md",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`CoreLibrary Unit Tests - PuppeteerPDFPrinter printManyPages 2`] = `
[MockFunction] {
  "calls": [
    [
      {
        "format": "a4",
        "path": "test\\test1.md.pdf",
      },
    ],
    [
      {
        "format": "a4",
        "path": "test\\test2.md.pdf",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`CoreLibrary Unit Tests - PuppeteerPDFPrinter printOnePages 1`] = `
[MockFunction] {
  "calls": [
    [
      "http://localhost:3000/test1.md",
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`CoreLibrary Unit Tests - PuppeteerPDFPrinter printOnePages 2`] = `
[MockFunction] {
  "calls": [
    [
      {
        "format": "a4",
        "path": "test\\test1.md.pdf",
      },
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
